### PR TITLE
Access control API docs

### PR DESCRIPTION
**Pull request by timleslie**
Preview deployment here: https://keystone-next-docs-git-access-control-api-keystonejs.vercel.app/apis/access-control

### CODE REVIEW

**Comment by gautamsi** [comment]
@timleslie I wanted to remind you of condition in #3239 which was fixed in #3241, there was one condition left that it may still have incorrect sequence if user has no access to update some items (based on declarative control like isLocked = true) then the sequence in the resolveInput and afterChange would be different in the array. 

I can explain more if it does not remind you full context.

**Comment by timleslie** [comment]
> @timleslie I wanted to remind you of condition in #3239 which was fixed in #3241, there was one condition left that it may still have incorrect sequence if user has no access to update some items (based on declarative control like isLocked = true) then the sequence in the resolveInput and afterChange would be different in the array.
> 
> I can explain more if it does not remind you full context.

Thanks for the reminder of these. I'm going to follow this up with a separate PR to fix that behaviour.

File: docs-next/pages/apis/access-control.mdx [hunk 1]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+          access: { /* ... */ },
+          /* ... */
+        }),
+      },
+      access: { /* ... */ },
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:25]:
```suggestion
        fieldName: text({
          access: { /* ... */ },
        }),
      },
      access: { /* ... */ },
    }),
  }),
```

File: docs-next/pages/apis/access-control.mdx [hunk 2]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+You can specify access control using either **concise** or **verbose** syntax.
+Verbose syntax provides a separate access control rule for each operation type; `create`, `read`, `update`, and `delete`.
+Concise syntax provides a single access control rule which is used for all operation types.
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:48]:
```suggestion
    ListKey: list({
```

afaik we call this the list **key** everywhere else

File: docs-next/pages/apis/access-control.mdx [hunk 3]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+        create: true,
+        read: true,
+        update: true,
+        delete: true,
+      },
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:62]:
```suggestion
    }),
  }),
```

these add no value

File: docs-next/pages/apis/access-control.mdx [hunk 4]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+export default config({
+  lists: createSchema({
+    ListName: list({
+      // Static access control definition
+      access: true,
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:95]:
```suggestion
      access: true,
    }),
  }),
```

File: docs-next/pages/apis/access-control.mdx [hunk 5]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+As such, Keystone will exclude the related operations and types from the GraphQL API.
+For example, if you set `{ create: false }` then the mutations `createItem` and `createItems` will be removed from the GraphQL API.
+If you want to keep the operations in the GraphQL API while preventing all access, you can use the [imperative](#imperative-list) access control rule `() => false`.
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:88]:
```suggestion
    ListKey: list({
```

File: docs-next/pages/apis/access-control.mdx [hunk 6]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+For update and delete operations, the access control rule is merged with the `id` value to form a filter.
+For singular operations, e.g. `updateItem` or `createItem`, if the merged filter does not return an item then the mutation will return an `Access Denied` error.
+For multi-item operations, e.g. `updateItems` or `createItems`, if the merged filter excludes items then these will simply be ignored by the operation, giving the same behaviour as if the ID was missing.
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:115]:
```suggestion
    ListKey: list({
```

File: docs-next/pages/apis/access-control.mdx [hunk 7]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+      fields: { isLocked: checkbox() },
+      // Declarative access control definition
+      access: { isLocked: false },
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:123]:
```suggestion
    }),
  }),
```

File: docs-next/pages/apis/access-control.mdx [hunk 8]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+If the function returns `false` then an `Access Denied` error will be returned.
+If the function returns `true` then the operation will be allowed.
+If the function returns a declarative value then this will be applied to the operation as decribed [above](#declarative-list).
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:149]:
```suggestion
    ListKey: list({
```

File: docs-next/pages/apis/access-control.mdx [hunk 9]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+      // Imperative access control definition
+      access: args => true,
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:156]:
```suggestion
    }),
  }),
```

File: docs-next/pages/apis/access-control.mdx [hunk 10]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+> **Important**: When writing imperative access control functions it is essential to consider both singular and multi-value operations.
+> If you check the `itemId` in the singular case, then you also need to check the `itemIds` in the multi-value case.
+> Similarly, `originalInput` will be different in the singular and multi-valued cases.
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:184]:
```suggestion
    ListKey: list({
```

File: docs-next/pages/apis/access-control.mdx [hunk 11]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+        itemIds,
+        context,
+      }) => {
+        return true;
+      },
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:202]:
```suggestion
    }),
  }),
```

File: docs-next/pages/apis/access-control.mdx [hunk 12]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+Verbose syntax provides a separate access control rule for each operation type; `create`, `read`, and `update`.
+Concise syntax provides a single access control rule which is used for all operation types.
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+import { text } from '@keystone-next/fields';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:223]:
```suggestion
    ListKey: list({
```

File: docs-next/pages/apis/access-control.mdx [hunk 13]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+          },
+          /* ... */
+        }),
+        /* ... */
+      },
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:245]:
```suggestion
          },
        }),
      },
    }),
  }),
```

File: docs-next/pages/apis/access-control.mdx [hunk 14]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+For example, if you set `{ update: false }` then the field would not appear in the `ItemUpdateInput` and `ItemsUpdateInputs` types of the GraphQL API.
+If you want to keep the fields in the GraphQL API while preventing all access, you can use the [imperative](#imperative-list) access control rule `() => false`.
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+import { text } from '@keystone-next/fields';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:272]:
```suggestion
    ListKey: list({
```

File: docs-next/pages/apis/access-control.mdx [hunk 15]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+          access: true,
+          /* ... */
+        }),
+        /* ... */
+      },
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:285]:
```suggestion
        }),
      },
    }),
  }),
```

File: docs-next/pages/apis/access-control.mdx [hunk 16]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+If the function returns `false` then an `Access Denied` error will be returned.
+If the function returns `true` then the operation will be allowed.
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+import { text } from '@keystone-next/fields';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:305]:
```suggestion
    ListKey: list({
```

File: docs-next/pages/apis/access-control.mdx [hunk 17]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+          access: args => true,
+          /* ... */
+        }),
+        /* ... */
+      },
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:318]:
```suggestion
        }),
      },
    }),
  }),
```

File: docs-next/pages/apis/access-control.mdx [hunk 18]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+| `context`       | The [`KeystoneContext`](./context) object of the originating GraphQL operation.                                                                               |
+| `item`          | The item being updated, deleted, or read. This object is an unresolved list item. See the [list item API](./list-items) for more details on unresolved list items. |
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+import { text } from '@keystone-next/fields';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:344]:
```suggestion
    ListKey: list({
```

File: docs-next/pages/apis/access-control.mdx [hunk 19]
```diff
@@ -2,6 +2,372 @@ import { Markdown } from '../../components/Page';
...
+          },
+          /* ... */
+        }),
+        /* ... */
+      },
+      /* ... */
+    }),
+    /* ... */
+  }),
+  /* ... */
```
Comment (JedWatson) [inline docs-next/pages/apis/access-control.mdx:369]:
```suggestion
        }),
      },
    }),
  }),
```

**Review by JedWatson — CHANGES_REQUESTED** [review]

**Review by JedWatson — APPROVED** [review]
Nice!

**Pull request merged**